### ISSUE TITLE
Fail to reproduce the paper result when training from scratch

### ISSUE THREAD

**Issue by didadida-r** [yxlu-0102/MP-SENet#15]
Hi,
    i try to train the model from scratch, but it fail to reproduce the result in VoiceBank_DEMAND, but the training is unstable and result is bad, can you give some advice.
    Thanks.

the tensorboard log is: 
![image](https://github.com/yxlu-0102/MP-SENet/assets/13691793/b5b4d094-42ac-4d3a-8a12-1f43c6a4f49e)

the only different is the config, and the diff result is:
```
 {
-    "batch_size": 4,
+    "batch_size": 8,
@@ -13,12 +13,12 @@
 
-    "segment_size": 32000,
+    "segment_size": 16000,
 
-    "num_workers": 4,
+    "num_workers": 8,
```

**Comment by yxlu-0102** [comment]
Since our model is based on self-attention, the segment size may have an impact on the overall performance, the tensorboard log should be like this:
<img width="1584" alt="截屏2024-01-07 12 43 01" src="https://github.com/yxlu-0102/MP-SENet/assets/82219664/45471857-ed71-461f-922c-341294e228ef">

**Comment by didadida-r** [comment]
Thank you for your response. I have adjusted the segment size to 32000. Here are the results from TensorBoard. After downloading the dataset and running the script, the training result is unstable, with a maximum PESQ of 3.3.

Is your tensorboard result[PESQ 3.56] reproduced from the offical setup config and code

![image](https://github.com/yxlu-0102/MP-SENet/assets/13691793/45083a94-abf5-43b5-8eb7-8fc25a983948)

**Comment by yxlu-0102** [comment]
This TensorBoard result is from our [subsequent improvements](https://arxiv.org/abs/2308.08926). 
I deleted the previous results, but there were no issues when I ran it before. 
From the tensorboard you provided, it seems there is a problem with the loss reduction in the magnitude spectrum.

**Comment by didadida-r** [comment]
Is the multi-GPU training process important? I am training the model using two GPUs.

**Comment by yxlu-0102** [comment]
The impact of multi-GPU training on the experimental results should be minimal.

**Comment by JangyeonKim** [comment]
![1](https://github.com/yxlu-0102/MP-SENet/assets/116805352/c61c9d3e-4e19-49ec-a39f-cde130aa6983)
![2](https://github.com/yxlu-0102/MP-SENet/assets/116805352/2ed1606e-1a81-417a-a9bd-a4d87b8c477e)
![3](https://github.com/yxlu-0102/MP-SENet/assets/116805352/0637db24-e768-49ec-a6d1-726242aad3e5)

Hello. Thank you for sharing the excellent code.

I am trying to replicate the performance on the DNS dataset after I read the long version and short version of MP-SENet but I am failing. I have matched all the configurations mentioned in the paper (2 sec segment, optimizer, lr, batch size, etc.). Could you provide some advice regarding this?

Additionally, for now, I am trying to see the performance of the generator without the Metric Discriminator. Could you share the loss graphs from the ablation study?

Any advice would be greatly appreciated.

**Comment by JangyeonKim** [comment]
`class MPNetLoss(nn.Module) :
    def __init__(self, h):
        super(MPNetLoss, self).__init__()
        self.mse_loss = nn.MSELoss()
        self.l1_loss = nn.L1Loss()
        self.h = h
        
    def forward(self, mag_pred, pha_pred, com_pred, S_true):
        clean_audio = S_true
        clean_mag, clean_pha, clean_com = mag_pha_stft(clean_audio, self.h)
        enhanced_mag, enhanced_pha, enhanced_com = mag_pred, pha_pred, com_pred
        enhanced_audio = mag_pha_istft(enhanced_mag, enhanced_pha, self.h)
        
        loss_mag = self.mse_loss(clean_mag, enhanced_mag)
        
        loss_ip, loss_gd, loss_iaf = phase_losses(clean_pha, enhanced_pha, self.h)
        loss_pha = loss_ip + loss_gd + loss_iaf
        
        loss_com = self.mse_loss(clean_com, enhanced_com) * 2
        
        loss_time = self.l1_loss(clean_audio, enhanced_audio)
        
        _, _, spec_for_concsistency = mag_pha_stft(enhanced_audio, self.h) # istft -> stft
        loss_con = self.mse_loss(enhanced_com, spec_for_concsistency) * 2
        
        # loss_all = 0.9 * loss_mag + 0.3 * loss_pha + 0.1 * loss_com + 0.2 * loss_time
        loss_all = 0.9 * loss_mag + 0.3 * loss_pha + 0.1 * loss_com + 0.1 * loss_con
        
        return {
            'magnitude_loss': loss_mag,
            'phase_loss': loss_pha,
            'complex_loss': loss_com,
            # 'time_loss': loss_time,
            'consistency_loss': loss_con,
            'total_loss': loss_all
        }`
        
 To aid in seeking advice, I am also attaching the code used to calculate the loss.

**Comment by yxlu-0102** [comment]
Sorry, I deleted the corresponding TensorBoard files after completing the
ablation experiment.

I reviewed your code, and it seems to be fine. Can you offer me the loss
curves of your results on the DNS dataset?

JangyeonKim ***@***.***> 于2024年5月24日周五 09:58写道：

**Comment by JangyeonKim** [comment]
[log.zip](https://github.com/yxlu-0102/MP-SENet/files/15427407/log.zip)

Thank you for your answer.
Are the loss curves you mentioned referring to the tensorboard logs? I have attached the file for your reference.

**Comment by yxlu-0102** [comment]
The loss curve does look quite strange. 
The phase loss during training doesn’t seem to decrease significantly, and both the magnitude and phase losses during validation are very odd. 

However, there is nothing wrong with the loss calculation code you provided. 
May I ask if the training was normal on the VoiceBank+DEMAND dataset?

**Comment by JangyeonKim** [comment]
I am currently training using only the DNS dataset.

I will download the 16kHz version of the VoiceBank+DEMAND dataset from your repository, apply it, and share the loss graph with you.

**Comment by yxlu-0102** [comment]
ok

**Comment by yunzqq** [comment]
> I am currently training using only the DNS dataset.
> 
> I will download the 16kHz version of the VoiceBank+DEMAND dataset from your repository, apply it, and share the loss graph with you.

If you can obtain the results for DEMAND in paper when you use the g_best model?

**Comment by JangyeonKim** [comment]
Oh, I forgot to report the experimental results. I am really sorry about that.
I only used the MP-SENet model and trained it in my training framework, and I have not used the g_best model.

Currently, the above issue has been resolved, and I have achieved a result of 3.41 WB-PESQ on the DNS dataset and 3.44 WB-PESQ on the VoiceBank+DEMAND dataset.

![image](https://github.com/yxlu-0102/MP-SENet/assets/116805352/4a105ed7-5f61-4f38-aa52-25f1e87564d1)

**Comment by JangyeonKim** [comment]
Additionally, I have some questions for the authors.

Currently, I am trying to apply the MP-SENet model to the BWE task. As written in the long version of the paper, I am conducting experiments with the VCTK dataset.

1. I changed the lsigmoid() of the mask decoder to prelu(), but the loss becomes nan as soon as the training starts. Leakyrelu() also showed the same phenomenon. So, I am currently using relu for training. Can you provide any advice regarding this issue?

`class MaskDecoder(nn.Module):
    def __init__(self, h, out_channel=1):
        super(MaskDecoder, self).__init__()
        self.dense_block = DenseBlock(h, depth=4)
        self.mask_conv = nn.Sequential(
            nn.ConvTranspose2d(h.dense_channel, h.dense_channel, (1, 3), (1, 2)),
            nn.Conv2d(h.dense_channel, out_channel, (1, 1)),
            nn.InstanceNorm2d(out_channel, affine=True),
            nn.PReLU(out_channel),
            nn.Conv2d(out_channel, out_channel, (1, 1))
        )
        self.lsigmoid = LearnableSigmoid_2d(h.n_fft//2+1, beta=h.beta)
        self.prelu = nn.PReLU()
    
    def forward(self, x):
        x = self.dense_block(x)
        x = self.mask_conv(x)
        x = x.permute(0, 3, 2, 1).squeeze(-1)
        
        # # lsigmoid for denoisig, dereverberation
        # x = self.lsigmoid(x).permute(0, 2, 1).unsqueeze(1)
        
        # PReLU for Bandwidth Extension
        x = self.prelu(x).permute(0, 2, 1).unsqueeze(1)
        
        return x`

2. When conducting experiments, aside from the metric scores, I found that the output samples contain audible artifacts (buzzing-like sound). I am curious if you have encountered the same issue.

**Comment by jeffery-work** [comment]
> Since our model is based on self-attention, the segment size may have an impact on the overall performance, the tensorboard log should be like this: <img alt="tensorboard" width="1578" src="https://private-user-images.githubusercontent.com/82219664/294738159-253b323e-7a28-4df2-b1fd-8ef2d10a4af7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE5NzI1MTYsIm5iZiI6MTcyMTk3MjIxNiwicGF0aCI6Ii84MjIxOTY2NC8yOTQ3MzgxNTktMjUzYjMyM2UtN2EyOC00ZGYyLWIxZmQtOGVmMmQxMGE0YWY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI2VDA1MzY1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAxNmU2NTk1NTYxMTVmMDdlYmQwZjhiZGYyNzE5N2FkOTJkZjcyMmE3ZTdmMzc0M2Y5NmNhNWE2YWEyYjg0ZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8361CTCvjLXc0qKzVcN5jpdLGECLdCeWT5iZRKfdR14">

So, if train segment is 2 seconds, then self-attention scope is 2 seconds? I check the AttentionModule implement, It seems like this. No wonder it takes me so long to inference once. Ha, ha

**Comment by xliu99** [comment]
> Oh, I forgot to report the experimental results. I am really sorry about that. I only used the MP-SENet model and trained it in my training framework, and I have not used the g_best model.
> 
> Currently, the above issue has been resolved, and I have achieved a result of 3.41 WB-PESQ on the DNS dataset and 3.44 WB-PESQ on the VoiceBank+DEMAND dataset.
> 
> ![image](https://private-user-images.githubusercontent.com/116805352/346580948-4a105ed7-5f61-4f38-aa52-25f1e87564d1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4ODYzMDYsIm5iZiI6MTcyODg4NjAwNiwicGF0aCI6Ii8xMTY4MDUzNTIvMzQ2NTgwOTQ4LTRhMTA1ZWQ3LTVmNjEtNGYzOC1hYTUyLTI1ZjFlODc1NjRkMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxNFQwNjA2NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05M2FkOTc1MjdkMjIzNDgyNmM3MzAzZDljNjQxZWRhYTIzYTcwNGM1MTliN2ZmOGY1MDk2MDU5Zjc0ODM4YmRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pxFbOw09SV_PXGY5K4t2tEsngHJpTu1YTMs_ho6TOcY)

Hi, may I know how you resolved the issue and made the training work?
